src/libraries/System.Private.CoreLib/src/System/IO/Path.cs

```diff
@@ -926,5 +926,56 @@ private static string GetRelativePath(string relativeTo, string path, StringComp
         /// Returns true if the path ends in a directory separator.
         /// </summary>
         public static bool EndsInDirectorySeparator(string path) => PathInternal.EndsInDirectorySeparator(path);
+
+        /// <summary>
+        /// Removes any redundant segments found in the specified path.
+        /// </summary>
+        /// <param name="path">A string containing an absolute or relative path that may or may not contain redundant segments.</param>
+        /// <returns><paramref name="path" /> without any redundant segments.</returns>
+        public static string RemoveRedundantSegments(string path)
+        {
+            if (path == null)
+                throw new ArgumentNullException(nameof(path));
+
+            if (PathInternal.IsEffectivelyEmpty(path))
+                throw new ArgumentException(SR.Arg_PathEmpty, nameof(path));
+
+            return RemoveRedundantSegments(path.AsSpan());
+        }
+
+        /// <summary>
+        /// Removes any redundant segments found in the specified path.
+        /// </summary>
+        /// <param name="path">A read-only span containing an absolute or relative path that may or may not contain redundant segments.</param>
+        /// <returns><paramref name="path" /> without any redundant segments.</returns>
+        public static string RemoveRedundantSegments(ReadOnlySpan<char> path)
+        {
+            if (PathInternal.IsEffectivelyEmpty(path))
+                throw new ArgumentException(SR.Arg_PathEmpty, nameof(path));
+
+            Span<char> destination = stackalloc char[path.Length];
```

Comment:
Do not have unbounded stackalloc based on user input. This could stackoverflow if the user provided a path that is too large.

Consider either using the arraypool or above a certain threshold just allocate a new array.

```C#
Span<char> destination = path.Length <= 256 ? stackalloc char[256] : new char[path.Length];
```

Reply:
Thanks for pointing that out, @ahsonkhan. Did you mean the condition to be `path.Length >= 256` ?

Reply:
Nope. You want to put an upper bound on the stack alloc for smaller/more common lengths.

Reply:
You might want to consider using `ValueStringBuilder` directly in the internal API- i.e. pass one in `PathInternal.TryRemoveRedundantSegments` instead of `Span`. I think it might make things cleaner.

Reply:
Is `ValueStringBuilder` necessary in this case or is path known to have a fixed upper bound? If the destination is always going to be <= the input, maybe we don't need the `Append`/`Grow` semantics of VSB outside the initial buffer.